bindings/rust/bench/src/s2n_tls.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    harness::{read_to_bytes, Mode, TlsBenchHarness},
+    CA_CERT_PATH, SERVER_CERT_CHAIN_PATH, SERVER_KEY_PATH,
+};
+use log::debug;
+use s2n_tls::{
+    callbacks::VerifyHostNameCallback,
+    config::{Builder, Config},
+    connection::Connection,
+    enums::Blinding,
+    security::DEFAULT_TLS13,
+};
+use std::{
+    collections::VecDeque,
+    ffi::c_void,
+    io::{Read, Write},
+    os::raw::c_int,
+    pin::Pin,
+    task::Poll::Ready,
+};
+
+pub struct S2nTls {
+    client_to_server_buf: Pin<Box<VecDeque<u8>>>, // Pinned pointer to a VecDeque to pass as C pointer for custom IO
+    server_to_client_buf: Pin<Box<VecDeque<u8>>>,
```

Comment:
This really should be an [`UnsafeCell`](https://doc.rust-lang.org/stable/std/cell/struct.UnsafeCell.html) since you have multiple pointers pointing to the values. Otherwise, rust assumes that `&mut` references cannot have any other references and makes some optimizations based on that, which can lead to undefined behavior

Reply:
I changed it to an `UnsafeCell<VecDeque<u8>>` which works. Do we need a `Pin<UnsafeCell<VecDeque<u8>>>` instead to guarantee the `VecDeque` doesn't move? Why or why not?

Reply:
I think to be safest, it's probably best to have a `Pin<Box<UnsafeCell<VecDeque<u8>>>>`. The `Pin<Box<T>>` put the memory on the heap, which makes it easier to hold a long-term pointer to it. And this struct is on the stack, which can technically move around. Then the `UnsafeCell<T>` is so we can tell rust that we know that there are multiple references to `T` and we are doing our own synchronization to ensure there aren't multiple `&mut` references to that underlying data.